functions/reacts.py

```diff
@@ -0,0 +1,156 @@
+
+def response_has_match_link(response: ReactResponse, match_id_results: set[str] | None = None) -> bool:
+    # check if the response is linked to any of the matches or if the reaction has other links to matches and is linked to any of the matched links
+    return not match_id_results or (response.match_link_id in match_id_results or any(link_id in match_id_results for link_id in response.other_match_link_ids))
+
+async def react_to_message(message: Message, possible_reactions: list[ReactPossibleReaction], match_id_results: set[str] = None) -> None:
+    for possible_reaction in possible_reactions:
+        # If the matched linked results exists check if the reaction is linked to a match
+        if not response_has_match_link(possible_reaction, match_id_results):
+            continue
+        # List of reactions in random order
+        reactions_list = sample(possible_reaction.reactions, len(possible_reaction.reactions))
+        reaction_count = 0
+        for reaction in reactions_list:
+            if possible_reaction.react_with_all: # If all of the possible reactions should be added
+                await add_reaction(message, reaction)
+                reaction_count += 1
+            elif possible_reaction.max_react_limit > 0 and reaction_count >= possible_reaction.max_react_limit:
+                break
+            else: # If the reaction should be added with a certain chance
+                if mock_bernoulli(possible_reaction.chance):
+                    await add_reaction(message, reaction)
+                    reaction_count += 1
+
+async def add_reaction(message: Message, reaction: str) -> None:
+    try:
+        await message.add_reaction(reaction)
+    except Exception as e:
+        print('Failed to react to message:', e, reaction)
+
+async def reply_to_message(message: Message, possible_replies: list[ReactPossibleReply], match_id_results: set[str] = None) -> None:
+    for possible_reply in possible_replies:
+        # If the matched linked results exists check if the reply is linked to a match
+        if not response_has_match_link(possible_reply, match_id_results):
+            continue
+        try:
+            if mock_bernoulli(possible_reply.chance):
+                # TODO: Handle different types of replies
+                await message.reply(possible_reply.message)
+        except Exception as e:
+            print('Failed to reply to message:', e, possible_reply)
+
+
+async def handle_message_react(message: Message)  -> list[str]:
+    events = []
+    for event_name, resource in REACT_RESOURCES.items():
+        match_results = check_matches(message.content, resource.criteria)
+        if bool(match_results):
+            # If the match results are not a set, it means that there was a match but no link IDs were found
+            # All the possible reactions/replies will be used in this case
+            if not isinstance(match_results, set):
+                match_results = None
```

Comment:
could return here right away because no matches?

Reply:
`match_link_id` is an optional ID, there can still be matches but if there's no link IDs available then all the possible reactions/replies will be used

Reply:
oh right